Fixes #33484 - add setting for opt-in/out the new host details

amirfefer opened the pull request:
<!---

Thank you for contributing to The Foreman, please read the
[following guide](https://www.theforeman.org/contribute.html), in short:

* [Create an issue](https://projects.theforeman.org/projects/foreman/issues)
* Reference the issue via `Fixes #1234` in the commit message
* Prefer present-tense, imperative-style commit messages
* Mark all strings for translation, see [1]
* Suggest prerequisites for testing and testing scenarios following example above.
* Prepend `[WIP]` for work in progress to prevent bots from triggering actions
* Be patient, we will do our best to take a look as soon as we can
* Explain the purpose of the PR, attach screenshots if applicable
* List all prerequisites for testing (e.g. VMware cluster, two smart proxies...)
* Reviewers often use extensive list of items to check, have a look prior submitting [2]
* Be nice and respectful

1: https://projects.theforeman.org/projects/foreman/wiki/Translating#Translating-for-developers
2: https://github.com/theforeman/foreman/blob/develop/developer_docs/pr_review.asciidoc
-->

amirfefer committed:
Fixes #33484 - add setting for opt-in/out the new host details

amirfefer commented on app/assets/javascripts/host_edit.js:
```diff
@@ -170,6 +170,9 @@ function submit_with_all_params() {
     data: serializeForm(),
     success: function(response) {
+      if (response.includes('layouts/react_application')) {
+        tfm.nav.pushUrl('/new/hosts/' + get_host_name());
```
this is for redirecting to the react show page after a successful edit/create

ezr-ondrej replied:
wouldn't the response contain the host id?

ekohl replied:
I'd like to note that hostnames are not by definition unique in Foreman. Hostnames can also change so I'd also say it should use the host ID

amirfefer replied:
the URL  for the legacy host show page uses the hostname, moreover `host_path` returns `/hosts/<fqdn>`
wouldn't it unique by definition?

ekohl replied:
I see it's that way now so I'm not going to object to this, but we have seen that it's possible to create multiple hosts with the same hostname (Foreman doesn't enforce this to be unique now). That's the background.

amirfefer commented:
This is a different approach from the previous attempt. 
the new host details page is going to replace the legacy one, so we need a simple mechanism for keeping these two temporarily. the previous PR solution is complex and has high coupling which might impact stability, maintenance, and unwanted leftovers after legacy page extraction.

ezr-ondrej commented on app/helpers/hosts_helper.rb:
```diff
@@ -433,4 +433,8 @@ def lookup_values_data(host_or_hostgroup)
     end
   end
+
+  def current_host_details_path(host)
```
You can use [`helper_method`](https://apidock.com/rails/ActionController/Helpers/ClassMethods/helper_method) in the controller and you'd not need this :)

ezr-ondrej reviewed (commented):
Make sense overall, let see what the tests say :)

ezr-ondrej commented:
[test katello]

ezr-ondrej commented:
Tests say we need the method in the helper directly :)

amirfefer commented:
[test foreman]

ezr-ondrej commented on app/assets/javascripts/host_edit.js:
```diff
@@ -170,6 +171,10 @@ function submit_with_all_params() {
     success: function(response) {
+      // workaround for redirecting to the new host details page
+      if (response.includes('layouts/react_application')) {
+        tfm.nav.pushUrl(`/new/hosts/${fqdn}`);
```
oh, we can't use backticks in the legacy JS, that's why the tests are failing.
```suggestion
        tfm.nav.pushUrl('/new/hosts/' + fqdn);
```

amirfefer replied:
right! I just got used to es6 :)

ezr-ondrej reviewed (commented):
LGTM :+1:
but found the issue CI is failing on.

ezr-ondrej commented on app/assets/javascripts/host_edit.js:
```diff
@@ -170,6 +171,10 @@ function submit_with_all_params() {
     success: function(response) {
+      // workaround for redirecting to the new host details page
+      if (response.includes('layouts/react_application')) {
+        tfm.nav.pushUrl('/new/hosts/' + fqdn);
```
This doesn't work for me. After host create, the page goes blank.
In the logs I see the request for React#index (on url `/new/hosts/<fqdn>`, so it does the correct thing, but nothing shows up, and I see white blank page with only menus visible.

Ron-Lavi replied:
@amirfefer it happens also on my end, blank page with this logs:
```bash
16:09:28 rails.1   | 2021-10-31T16:09:28 [I|app|58ffdc76] Started GET "/new/hosts/linda-gechas.example.com" for 10.35.206.105 at 2021-10-31 16:09:28 +0200
16:09:28 rails.1   | 2021-10-31T16:09:28 [I|app|58ffdc76] Processing by ReactController#index as */*
16:09:28 rails.1   | 2021-10-31T16:09:28 [I|app|58ffdc76]   Parameters: {"id"=>"linda-gechas.example.com"}
16:09:28 rails.1   | 2021-10-31T16:09:28 [I|app|58ffdc76]   Rendering react/index.html.erb within layouts/react_application
16:09:28 rails.1   | 2021-10-31T16:09:28 [I|app|58ffdc76]   Rendered react/index.html.erb within layouts/react_application (Duration: 0.1ms | Allocations: 37)
16:09:28 rails.1   | 2021-10-31T16:09:28 [I|app|58ffdc76]   Rendering layouts/base.html.erb
16:09:28 rails.1   | 2021-10-31T16:09:28 [I|app|58ffdc76]   Rendered layouts/base.html.erb (Duration: 309.2ms | Allocations: 80333)
16:09:28 rails.1   | 2021-10-31T16:09:28 [I|app|58ffdc76] Completed 200 OK in 320ms (Views: 308.7ms | ActiveRecord: 4.0ms | Allocations: 84773)

```

amirfefer replied:
Thanks, logs seem ok, is there any console error?  it works fine on my env

ezr-ondrej replied:
There is no console error, but the response the `if` returns false for me, so it fails to recognize that this was the new page response.
Here what is the `response` is for me (from debugging): https://pastebin.com/UZvUtSwG

Looks quite empty to me :thinking:

ezr-ondrej replied:
Could we expose the setting through some data attribute on the form? that should not be hard I'd guess and probably bit more relyable?

ezr-ondrej reviewed (commented):
The host create doesn't work for me. Apart of that, everything looks working fine :+1:

amirfefer commented:
works for me now:

https://user-images.githubusercontent.com/11807069/139589411-5dab91ac-5650-4c6c-a8a4-da701e373f3e.mov


rebased and added a few intergration tests

ezr-ondrej commented:
Unfortunatelly it does not work for me...

https://user-images.githubusercontent.com/2884324/139595258-76f604b6-0646-4d20-b0c3-789cf3f6c6c4.mp4

EDIT: here is the response that I get with the new host details page on https://pastebin.com/UZvUtSwG

amirfefer commented:
@ezr-ondrej, it should work now, the test has passed :)

ezr-ondrej reviewed (approved):
Thanks @amirfefer ! Works for me, will wait for the CI.

ezr-ondrej commented on app/assets/javascripts/host_edit.js:
```diff
@@ -159,6 +159,7 @@ function update_capabilities(capabilities) {
 var stop_pooling;
 
 function submit_with_all_params() {
+  var fqdn = $('.fqdn')[0].textContent;
```
This is wrong, I've been reading it incorrectly and it is quite nasty one: https://projects.theforeman.org/issues/34300.
